### PR TITLE
5.x - Move primary term from ReplicationRequest to ConcreteShardRequest

### PR DESCRIPTION
**Pull request by ywelsch**
This is the 5.x equivalent to #25822

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 1]
```diff
@@ -1017,24 +1041,29 @@ public void onFailure(Exception shardFailedError) {
...
         }
 
-        ConcreteShardRequest(R request, String targetAllocationID) {
+        ConcreteShardRequest(R request, String targetAllocationID, long primaryTerm) {
             Objects.requireNonNull(request);
             Objects.requireNonNull(targetAllocationID);
             this.request = request;
             this.targetAllocationID = targetAllocationID;
+            this.primaryTerm = primaryTerm;
+            this.request.primaryTerm(primaryTerm); // for bwc with ES < v5.6, that has the primary term on the inner ReplicationRequest
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: can we add an assertion to replication request that allows to set it once to a non zero value and makes sure that future callers keep it the same?

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 2]
```diff
@@ -681,7 +694,9 @@ private void performLocalAction(ClusterState state, ShardRouting primary, Discov
                 logger.trace("send action [{}] on primary [{}] for request [{}] with cluster state version [{}] to [{}] ",
                     transportPrimaryAction, request.shardId(), request, state.version(), primary.currentNodeId());
             }
-            performAction(node, transportPrimaryAction, true, new ConcreteShardRequest<>(request, primary.allocationId().getId()));
+            final long primaryTerm = state.metaData().getIndexSafe(primary.index()).primaryTerm(primary.id());
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: nit: we can pass the indexMetaData (or primary term) as an argument. We look it up on in the calling method. Save on double lookups.

File: core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java [hunk 3]
```diff
@@ -823,15 +829,21 @@ public void testDefaultWaitForActiveShardsUsesIndexSetting() throws Exception {
...
+    /** test that a primary request is rejected if it arrives at a shard with a wrong allocation id or term */
+    public void testPrimaryActionRejectsWrongAidOrWrongTerm() throws Exception {
         final String index = "test";
         final ShardId shardId = new ShardId(index, "_na_", 0);
         setState(clusterService, state(index, true, ShardRoutingState.STARTED));
+        final ShardRouting primary = clusterService.state().routingTable().shardRoutingTable(shardId).primaryShard();
+        final long primaryTerm = clusterService.state().metaData().index(shardId.getIndexName()).primaryTerm(shardId.id());
         PlainActionFuture<Response> listener = new PlainActionFuture<>();
+        final boolean wrongAllocationId = randomBoolean();
+        final long wrongTerm = primaryTerm + randomIntBetween(1, 10);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java]: nit: same comment as on master - allow for having the right term and the wrong aid.

File: core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java [hunk 4]
```diff
@@ -171,12 +172,12 @@ long routedBasedOnClusterVersion() {
...
     /** returns the primary term active at the time the operation was performed on the primary shard */
-    public long primaryTerm() {
+    long primaryTerm() {
         return primaryTerm;
     }
 
     /** marks the primary term in which the operation was performed */
-    public void primaryTerm(long term) {
+    void primaryTerm(long term) {
         primaryTerm = term;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java]: can we assert we can only set it if it's zero and if it's not that it's equal to the parameter?

File: core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java [hunk 5]
```diff
@@ -171,12 +172,12 @@ long routedBasedOnClusterVersion() {
     }
 
     /** returns the primary term active at the time the operation was performed on the primary shard */
-    public long primaryTerm() {
+    long primaryTerm() {
         return primaryTerm;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java]: can we assert it's non-0? I worried now that people will use this method while we have the true primary term somewhere. Maybe also adapt the java docs and deprecate the method?

File: core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java (hunk 4 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java]: good call on making it package private.

**Review by bleskes — APPROVED** [review]
LGTM. I left some asks for extra assertions.

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java (hunk 1 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: good idea. I don't think that this holds though in the presence of retries that bubble up to the reroute phase. We could set it back to 0 though in the retry method of the reroute phase.

File: core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java (hunk 5 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java]: we cannot assert "non-0" here, as we are calling the method when deserializing a ConcreteShardRequest which comes from a < 5.6 node. Such a node will not be sending a "non-0" value for the primary action, hence the current special handling in `PrimaryOperationTransportHandler`

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java (hunk 1 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: I pushed 2306a3e which implements my adaption of your suggestion.

**Comment by ywelsch** [comment]
Thanks @bleskes

**Pull request merged**